src/repr/src/row/encode.rs

```diff
@@ -1239,6 +1239,24 @@ pub struct RowColumnarDecoder {
     nullability: Option<NullBuffer>,
 }
 
+/// Merge the provided null buffer with the existing array's null buffer, if any.
+fn mask_nulls(column: &ArrayRef, null_mask: Option<&NullBuffer>) -> ArrayRef {
+    if null_mask.is_none() {
+        Arc::clone(column)
+    } else {
+        // We calculate stats on the nested arrays, so make sure we don't count entries
+        // that are masked off at a higher level.
+        let nulls = NullBuffer::union(null_mask, column.nulls());
+        let data = column
+            .to_data()
+            .into_builder()
+            .nulls(nulls)
+            .build()
+            .expect("changed only null mask");
```

Comment:
maybe feels a little bad that we have to convert into `ArrayData` and back, but the runtime penalty here shouldn't be bad?

Reply:
It's not great - deeply nested structs will pay the cost multiple times. I think it's probably acceptable vs. the bug, though?

I intend to see if I can follow up with a nicer version of this stuff, but I expect that to be more work...

Reply:
> I think it's probably acceptable vs. the bug, though?

I agree!